# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/fs/parent/datamodel/FileSystemDataManager.cpp
# Commit: 75dd8ef65751
# Full Hash: 75dd8ef6575143b27c8479af06c86bbe4610e0df
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2022-09-05 21:43:29
# Description:
#   Bug 1789075 - Call RequestAllowToClose on correct thread; r=dom-storage-reviewers,jari
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D156394
# ==============================================================================

diff -r db3dfa11a0ec -r 75dd8ef65751 dom/fs/parent/datamodel/FileSystemDataManager.cpp
--- a/dom/fs/parent/datamodel/FileSystemDataManager.cpp	Mon Sep 05 11:58:59 2022 +0000
+++ b/dom/fs/parent/datamodel/FileSystemDataManager.cpp	Mon Sep 05 12:01:25 2022 +0000
@@ -298,24 +298,10 @@
     // can't be blocked by this Manager if there is no acquired DirectoryLock.
     // If there is an acquired DirectoryLock, check if the table contains the
     // lock for the Manager.
-    if (!quota::Client::IsLockForObjectAcquiredAndContainedInLockTable(
+    if (quota::Client::IsLockForObjectAcquiredAndContainedInLockTable(
             *dataManager, aDirectoryLockIds)) {
-      continue;
+      dataManager->RequestAllowToClose();
     }
-
-    InvokeAsync(dataManager->MutableIOTargetPtr(), __func__,
-                [dataManager = RefPtr<FileSystemDataManager>(
-                     dataManager.get())]() mutable {
-                  dataManager->RequestAllowToClose();
-
-                  nsCOMPtr<nsISerialEventTarget> target =
-                      dataManager->MutableBackgroundTargetPtr();
-
-                  NS_ProxyRelease("ReleaseFileSystemDataManager", target,
-                                  dataManager.forget());
-
-                  return BoolPromise::CreateAndResolve(true, __func__);
-                });
   }
 }
 
@@ -328,19 +314,7 @@
   }
 
   for (const auto& dataManager : gDataManagers->Values()) {
-    InvokeAsync(dataManager->MutableIOTargetPtr(), __func__,
-                [dataManager = RefPtr<FileSystemDataManager>(
-                     dataManager.get())]() mutable {
-                  dataManager->RequestAllowToClose();
-
-                  nsCOMPtr<nsISerialEventTarget> target =
-                      dataManager->MutableBackgroundTargetPtr();
-
-                  NS_ProxyRelease("ReleaseFileSystemDataManager", target,
-                                  dataManager.forget());
-
-                  return BoolPromise::CreateAndResolve(true, __func__);
-                });
+    dataManager->RequestAllowToClose();
   }
 }
 
@@ -365,16 +339,16 @@
 
 void FileSystemDataManager::RegisterActor(
     NotNull<FileSystemManagerParent*> aActor) {
-  MOZ_ASSERT(!mActors.Contains(aActor));
+  MOZ_ASSERT(!mBackgroundThreadAccessible.Access()->mActors.Contains(aActor));
 
-  mActors.Insert(aActor);
+  mBackgroundThreadAccessible.Access()->mActors.Insert(aActor);
 }
 
 void FileSystemDataManager::UnregisterActor(
     NotNull<FileSystemManagerParent*> aActor) {
-  MOZ_ASSERT(mActors.Contains(aActor));
+  MOZ_ASSERT(mBackgroundThreadAccessible.Access()->mActors.Contains(aActor));
 
-  mActors.Remove(aActor);
+  mBackgroundThreadAccessible.Access()->mActors.Remove(aActor);
 
   if (IsInactive()) {
     BeginClose();
@@ -394,11 +368,11 @@
 }
 
 bool FileSystemDataManager::IsInactive() const {
-  return !mRegCount && !mActors.Count();
+  return !mRegCount && !mBackgroundThreadAccessible.Access()->mActors.Count();
 }
 
 void FileSystemDataManager::RequestAllowToClose() {
-  for (const auto& actor : mActors) {
+  for (const auto& actor : mBackgroundThreadAccessible.Access()->mActors) {
     actor->RequestAllowToClose();
   }
 }